source/app/models/url.rb

```diff
@@ -0,0 +1,41 @@
+require 'uri'
+require 'net/http'
+
+class Url < ActiveRecord::Base
+  ALPHANUMERIC_CHARACTERS = (('a'..'z').to_a + ('A'..'Z').to_a + ('0'..'9').to_a)
+
+  validates_presence_of :original_url
+  validates_format_of :original_url, :with => URI::regexp, :message => "must be a valid Ruby URI"
+  validate :original_url_starts_with_http_or_https
+  validate :original_url_must_return_a_response
+
+  before_save :shorten_url
+
+  def original_url_starts_with_http_or_https
+    if not original_url =~ /^https?:\/\//
+      errors.add(:original_url, "must start with http:// or https://")
+    end
+  end
+
+  def original_url_must_return_a_response
+    return unless original_url.present? && original_url =~ URI::regexp(['http', 'https'])
```

Comment:
Matching the url against `URI`'s regexp helper gizmo is somewhat duplicative with your check in `original_url_starts_with_http_or_https`. I wonder how we can remove the duplicated concept. 🤔